### ISSUE TITLE
Binder looses binderNode.defaultValue for nested keys inside repeat() in Hilla 2.2.0

### ISSUE THREAD

**Issue by KardonskyRoman** [vaadin/hilla#1294]
### Description of the bug 
Binder looses binderNode.defaultValue for nested keys inside repeat() when rendering array model
### Minimal reproducible example
The array model passed to Binder:
```

[
    {
        "organization": {
            "address": {
                "building": "",
                "city": "",
                "country": "",
                "district": "",
                "note": "",
                "region": "",
                "room": "",
                "settlement": "",
                "street": ""
            }
        }
    }
]

```
This is how binder is initialized:
```

const declarant = TDeclarantModel.createEmptyValue();
declarant.organization = TOrganizationModel.createEmptyValue();
declarant.organization.address = TAddressModel.createEmptyValue();
this.binder.value.projectDocumentsDeveloper = [declarant];

```
If BinderNode has a nested keys ('address' in example), default values of nested keys are undefined.
### Expected behavior
BinderNode.defaultValue should looks like:

```
{
    "organization": {
        "address": {
            "building": "",
            "city": "",
            "country": "",
            "district": "",
            "note": "",
            "region": "",
            "room": "",
            "settlement": "",
            "street": ""
        }
    }
} 

```
### Actual behavior
The BinderNode.defaultValus is:

{
      organization: undefined
}

However, BinderNode.value is correct ('organization' is an object as expected)

So, I get en error:
```

Encountered an uncaught exception that was thrown by a reaction or observer component, in: 'Reaction[ConclusionView.update()]' TypeError: Cannot read properties of undefined (reading 'address')
    at get defaultValue [as defaultValue] (BinderNode.ts:153:26)
    at g2.initializeValue (BinderNode.ts:422:73)
    at new g2 (BinderNode.ts:94:10)
    at b2 (Models.ts:23:26) 
```

### Versions:
    - Vaadin / Hilla version: 2.2.0
    - Node version: 18.7
    - Java version: 17
    - OS version: Windows 10
    - Browser version (if applicable): Chrome
    - Application Server (if applicable):
    - IDE/Editor (if applicable):

**Comment by KardonskyRoman** [comment]
I noticed in 2.2, will try in 2.1.10

**Comment by KardonskyRoman** [comment]
Tried to set hilla version to 2.1.10, but frontend generation fails:
![image](https://github.com/vaadin/hilla/assets/39597003/8f837467-e4dd-4256-866c-12f369e51236)

**Comment by KardonskyRoman** [comment]
> Does this work in 2.1 or did you just notice it in 2.2?

Hilla 2.1.10 shows the same result.

**Comment by platosha** [comment]
Nested keys initialisation is lazy by design.

Creating a BinderNode for a particular nested field should trigger initialisation for `defaultValue` and `value` simultaneously. Could you try accessing the nested keys using the `binder.for` and models API?

```typescript
binder.for(binder.model.organization.address).defaultValue;
```

Another related suggestion: setting `binder.value` does not change `binder.defaultValue`, the difference indicates that the form is dirty (changed by the user). If you initialise the form, please try using `binder.read(value);` that sets `value` and `defaultValue` to the same argument value.

**Comment by platosha** [comment]
In theory, you should not need to call `createEmptyValue()` manually for nested things, for as long as you always access the nested things using models and `binder.for(binder.model.some.thing);` (or `field(model.some.thing)` in the form template).

**Comment by KardonskyRoman** [comment]
> In theory, you should not need to call `createEmptyValue()` manually for nested things, for as long as you always access the nested things using models and `binder.for(binder.model.some.thing);` (or `field(model.some.thing)` in the form template).

I called `createEmptyValue()` manually for nested key for testing purpose trying to solve the problem.

**Comment by KardonskyRoman** [comment]
> Nested keys initialisation is lazy by design.
> 
> Creating a BinderNode for a particular nested field should trigger initialisation for `defaultValue` and `value` simultaneously. Could you try accessing the nested keys using the `binder.for` and models API?
> 
> ```ts
> binder.for(binder.model.organization.address).defaultValue;
> ```
> 
> Another related suggestion: setting `binder.value` does not change `binder.defaultValue`, the difference indicates that the form is dirty (changed by the user). If you initialise the form, please try using `binder.read(value);` that sets `value` and `defaultValue` to the same argument value.

This is how I initialize the Binder:
`this.binder.clear();
    const conclusion = ConclusionModel.createEmptyValue();
    const documents = TDocumentsModel.createEmptyValue();
    const document = TDocumentModel.createEmptyValue();
    documents.document = [document];
    conclusion.documents = documents;
    const declarant = TDeclarantModel.createEmptyValue();
    declarant.organization = TOrganizationModel.createEmptyValue();
    declarant.organization.address = TAddressModel.createEmptyValue();
    declarant.organization.address.country = "";
    declarant.organization.address.building = "";
    declarant.organization.address.city = "";
    declarant.organization.address.district = "";
    declarant.organization.address.note = "";
    declarant.organization.address.region = "";
    declarant.organization.address.room = "";
    declarant.organization.address.settlement = "";
    declarant.organization.address.street = "";
    conclusion.projectDocumentsDeveloper = [declarant];
    this.binder.read(conclusion);`

And the value and default value of the binder now is correct.

The default value is not correct only for BinderNode

**Comment by KardonskyRoman** [comment]
The value of `this.binder.for(declarant.model.organization).defaultValue`  inside of `repeat( this.binder.model.projectDocumentsDeveloper,
                  (declarant) => {}` is undefined, so as `this.binder.for(declarant.model.organization.address).defaultValue` throws an error that too

**Comment by platosha** [comment]
As a workaround, would it help to call `binderNode.initializeValue(true)`? For example:

```typescript
const addressBinderNode = this.binder.for(declarant.model.organization.address).initializeValue(true);
addressBinderNode.defaultValue
```

**Comment by KardonskyRoman** [comment]
Property 'initializeValue' is private and only accessible within class 'BinderNode<T, M>'


> As a workaround, would it help to call `binderNode.initializeValue(true)`? For example:
> 
> ```ts
> const addressBinderNode = this.binder.for(declarant.model.organization.address);
> addressBinderNode.initializeValue(true);
> addressBinderNode.defaultValue
> ```

**Comment by platosha** [comment]
Should be pubic in `2.2.0`. Or you could add `// @ts-ignore` to suppress this error just to try it quickly.

**Comment by KardonskyRoman** [comment]
> Property 'initializeValue' is private and only accessible within class 'BinderNode<T, M>'
> 
> > As a workaround, would it help to call `binderNode.initializeValue(true)`? For example:
> > ```ts
> > const addressBinderNode = this.binder.for(declarant.model.organization.address);
> > addressBinderNode.initializeValue(true);
> > addressBinderNode.defaultValue
> > ```

Still throws  same error, tried 
`
const organizationBinderNode = this.binder.for(declarant.model.organization);
organizationBinderNode.initializeValue(true);
console.log(organizationBinderNode.defaultValue)
`
it prints `undefined`

**Comment by platosha** [comment]
Thanks a lot for the detailed info and for trying out the suggestions so far. That gave me some clues for the source of this issue.

Seems that `BinderNode.initializeValue()` does not verify that `defaultValue` gets initialised in case when the `value` was already set. We should probably add a check for `defaultValue` presence for this case.

**Comment by KardonskyRoman** [comment]
Thank you, too. Make Hilla work again!

**Comment by KardonskyRoman** [comment]
Due to there is no workaround for this bug, I would really appreciate if you make the ticket more critical, thanks.

**Comment by cromoteca** [comment]
I tried to reproduce, but I couldn't. Could you provide a minimal reproducible example?

**Comment by cromoteca** [comment]
Thanks for your effort. Turns out I couldn't reproduce because I was missing a level of nesting. The binder is able to find values correctly in, say, `personBinder.model.address.country,` but not `personBinder.model.organization.address.country`. I confirm the issue and I reproduced it in Hilla 2.2 and 1.3, just to make sure it's not a regression.

And, as you pointed out, the issue is not related to `createEmptyValue`, but also happens when dealing with full valid data coming from the server.
